test/integration_test.go

```diff
@@ -1819,6 +1819,82 @@ func (ts *IntegrationTestSuite) TestTallyScopeAccess() {
 	assertHistDuration("some_histogram", 5*time.Second, 2)
 }
 
+func (ts *IntegrationTestSuite) TestReturnCancelError() {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	wfIDPrefix := "test-return-cancel-error-"
+
+	// For most tests we don't return the raw error since it loses context
+	rawActivityError := false
+
+	// Activity using temporal canceled error when not canceled should return
+	// "unexpected activity cancel error"
+	fromActivity, waitForCancel, goCancelError := true, false, false
+	err := ts.executeWorkflow(wfIDPrefix+"1", ts.workflows.ReturnCancelError, nil,
+		fromActivity, rawActivityError, waitForCancel, goCancelError)
+	ts.Error(err)
+	ts.Contains(err.Error(), "unexpected activity cancel error")
+
+	// Activity using Go canceled error when not canceled should return a context
+	// canceled error
+	fromActivity, waitForCancel, goCancelError = true, false, true
+	err = ts.executeWorkflow(wfIDPrefix+"2", ts.workflows.ReturnCancelError, nil,
+		fromActivity, rawActivityError, waitForCancel, goCancelError)
+	ts.Error(err)
+	ts.Contains(err.Error(), "context canceled")
+
+	// Activity using temporal canceled error after cancel should return normal
+	// cancel error
+	fromActivity, waitForCancel, goCancelError = true, true, false
+	err = ts.executeWorkflow(wfIDPrefix+"3", ts.workflows.ReturnCancelError, nil,
+		fromActivity, rawActivityError, waitForCancel, goCancelError)
+	ts.Error(err)
+	ts.NotContains(err.Error(), "unexpected")
+	ts.Contains(err.Error(), "canceled")
+	// We also check that, since rawActivityError is false, this is _not_ a
+	// canceled workflow since just the error string is used. This assertion is
+	// only made here to show it's the opposite of the raw one later.
+	ts.False(temporal.IsCanceledError(err))
+	resp, err := ts.client.DescribeWorkflowExecution(ctx, wfIDPrefix+"3", "")
+	ts.NoError(err)
+	ts.Equal(enumspb.WORKFLOW_EXECUTION_STATUS_FAILED, resp.GetWorkflowExecutionInfo().GetStatus())
+
+	// Activity using Go canceled error after cancel should return normal cancel
+	// error
+	fromActivity, waitForCancel, goCancelError = true, true, true
+	err = ts.executeWorkflow(wfIDPrefix+"4", ts.workflows.ReturnCancelError, nil,
+		fromActivity, rawActivityError, waitForCancel, goCancelError)
+	ts.Error(err)
+	ts.NotContains(err.Error(), "context canceled")
+	ts.Contains(err.Error(), "canceled")
```

Comment:
add assert that it does not contain "unexpected"

Reply:
Unexpected only relates to temporal canceled errors not go canceled errors. Notice I don't have an assert that temporal canceled errors don't contain "context canceled". They are separate things.

No harm in adding it I guess.

Reply:
Added